### DISCUSSION TITLE
C++ image array always shows image shifted right

### DISCUSSION THREAD

**Discussion by gittyupnow** [ImageMagick/ImageMagick#6747] [Help] [answered]
### ImageMagick version

7.1.1-18 Q16-HDRI x86_64 

### Operating system, version and so on

Ubuntu 22.04 (LTS) x64

### Description

All my conversions to .h C++ image arrays are resulting in the image being shifted over to the right then the far right side of the image showing up on the left to fill filling the void. 

![eink-shift](https://github.com/ImageMagick/ImageMagick/assets/3425916/2ca21cfa-d88d-4b18-98f9-d2222f0c6d28)

Those screenshots are from [image2cpp](https://javl.github.io/image2cpp/) , but it looks the exact same on my eink screen.  Note that it does work if I use that site and the sample eink images (rather than imagemagick to convert an image, so I know it's not something wrong with my eink code itself).

Here are the commands I'm using:

```
./magick convert wizard: -resize '400x300!' -ordered-dither o2x2 -remap 4gray.png -extent 4:3 -depth 16 wizardout.bmp

./magick wizardout.bmp -depth 1 -negate wizardarray.h
```

When comparing a fully working array (top) to my imagemagick generated one (bottom) I notice two things:

![wiz-array-compare](https://github.com/ImageMagick/ImageMagick/assets/3425916/3f2dd6fa-c265-4314-a8de-f465c55979e9)

1) The working arrays have a size of 15000.  The imagemagik ones are 15011.
2) The working arrays have 16 hex values per line.  The imagemagik ones have 12.

Anyone know what I'm doing wrong?  Thank you,

-Mike

**Comment by snibgo** [comment]
> ./magick convert wizard: -resize '400x300!' ... etc

I suggest using just `magick`, not `magick convert`.

You don't supply `4gray.png`, so I can't reproduce your work.

I'm not sure what result you want. Perhaps you want an array of pixels in the `.h` file. IM gives you slightly more than that. See the top of the `.h` file:
```
/*
  wizardarray.h (PNM).
*/
static const unsigned char
  MagickImage[] =
  {
    0x50, 0x36, 0x0A, 0x34, 0x30, 0x30, 0x20, 0x33, 0x30, 0x30, 0x0A, 0x31, 
    0x0A, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 
```

The format is `PNM`, which has a small header. See https://en.wikipedia.org/wiki/Netpbm

> 2. The working arrays have 16 hex values per line. The imagemagik ones have 12.

That is simply a difference to where the line-breaks occur.

**Reply by gittyupnow** [reply]
Hey, thanks a lot for the quick reply.  Your tip regarding the included header fixed it for me! If I delete the first 11 values in the array it works perfectly (no image shift).  Is there a way to get ImageMagick to save without the header, or should I just remove it programmatically on my own?

**Comment by snibgo** [comment, marked as answer]
The header isn't a fixed 11 bytes. See the link I supplied. The header is text, which you can read to get the image dimensions etc. In the `.h` format, you can't turn off the header.

ImageMagick has alternative text output formats, such as `ftxt:`. See [ftxt: formatted text](http://im.snibgo.com/fmttxt.htm)
```
magick rose: -fx "u/QuantumRange*255" -define ftxt:format=\h,\n ftxt:
```
The output is headerless, one line per pixel, like this:
```
#47,#60,#35,
#47,#62,#3c,
#3a,#4d,#38,
#40,#52,#3a,
#48,#5c,#41,
#34,#42,#31,
```
You could pipe through `tr` or `sed` to change each `#` to `0x`, etc.

EDIT: Corrected "256" to "255".

**Reply by gittyupnow** [reply]
Thank you very much!  This is exactly what I needed.

**Reply by snibgo** [reply]
Sorry, I made an error. See my correction above.
